natmod: Allow linking static libraries.

agatti opened the pull request:
### Summary

This is an update to #15838 - as requested in https://github.com/micropython/micropython/pull/15838#issuecomment-2696048612.

All the good stuff is from @vshymanskyy, the not-so-good stuff is mine:

* Addressed the review points raised in the original PR (cache salt key, linker script detection logic, long argument for -l, some text changes, ESP-IDF CI caching)
* Picolibc is now supported for RV32 via a rather nasty workaround
* Xtensa natmods are now built as part of the CI process
* Natmods have the appropriate platform exceptions in the CI build step to make it work out of the box for most platforms
* The object cache directory is removed as part of `make clean`
* The object cache directory now contains a `CACHEDIR.TAG` file, so most archival/backup programs will skip that directory (if rust can do it why can't we?)
* Removed the assumption that there's a `python` executable in the CI environment (minor regression from 8b1ed4473dd215d564498f577cb6f3339c3c2e57), blocking manual testing on my provisioned build VMs :)

### Testing

All natmod-related CI jobs ran without issues on my own repo's branch.

vshymanskyy committed:
tools/mpy_ld.py: Allow linking static libraries.

This commit introduces an additional symbol resolution mechanism to the
natmod linking process.  This allows the build scripts to look for required
symbols into selected libraries that are provided by the compiler
installation (libgcc and libm at the moment).

For example, using soft-float code in natmods, whilst technically possible,
was not an easy process and required some additional work to pull it off.
With this addition all the manual (and error-prone) operations have been
automated and folded into `tools/mpy_ld.py`.

Both newlib and picolibc toolchains are supported, albeit the latter may
require a bit of extra configuration depending on the environment the build
process runs on.  Picolibc's soft-float functions aren't in libm - in fact
the shipped libm is nothing but a stub - but they are inside libc.  This is
usually not a problem as these changes cater for that configuration quirk,
but on certain compilers the include paths used to find libraries in may
not be updated to take Picolibc's library directory into account.  The bare
metal RISC-V compiler shipped with the CI OS image (GCC 10.2.0 on Ubuntu
22.04LTS) happens to exhibit this very problem.

To work around that for CI builds, the Picolibc libraries' path is
hardcoded in the Makefile directives used by the linker, but this can be
changed by setting the PICOLIBC_ROOT environment library when building
natmods.

Signed-off-by: Volodymyr Shymanskyy <vshymanskyi@gmail.com>
Co-authored-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
For the ESP32 target, successful build run logs can be found here https://github.com/agatti/micropython/actions/runs/13666850878/job/38209736600 and here https://github.com/agatti/micropython/actions/runs/13666850878/job/38209737465

The commit checker fails to see the sign-off line, but I guess fixing that is its own PR.

agatti committed:
tools/ci.sh: Build Xtensa natmods as part of the CI process.

This commit expands the CI tests by checking whether the example native
modules are able to be built for the Xtensa architecture.

This was made possible by the changes to mpy_ld that allow symbol
resolution across standard compiler-provided libraries.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tools/ci.sh: Do not assume the Python interpreter is called "python".

This commit removes the assumption made by the CI scripts that the
system-provided python executable is simply named "python".  The scripts
will now look for a binary called "python3" first, and then fall back to
"python" if that is not found.

Whilst this is currently the case for the CI environment, there are no
guarantees for this going forward.  For example minimal CI environments
set up by some developers, using the same base OS, have their python
executable called "python3".

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented on tools/ar_util.py:
```diff
@@ -0,0 +1,239 @@
+    if not is_ld_script():
+        return [CachedArFile(fn)]
+    else:
+        return [CachedArFile(item) for item in expand_ld_script(fn)]
```
I see you inverted the logic here by swapping the if/else.  But I think @projectgus had a different idea in mind in his comment https://github.com/micropython/micropython/pull/15838#discussion_r1942090278

I think what was intended was:
```py
ar_header = "!<arch>\012"
with open(fn, "rb") as f:
    is_ar_file = f.read(len(ar_header)) == ar_header

if is_ar_file:
    return [CachedArFile(fn)]
else:
    return [CachedArFile(item) for item in expand_ld_script(fn)]
```

agatti replied:
I see that now - I'm on it.  I was in a bit of a hurry back then to get the RV32 stuff working (I assumed it worked out of the box like when I tried it last time).

dpgeorge reviewed (commented):
Thanks @agatti for putting this PR together and tidying up the final bits.  I have just one comment above, but otherwise this looks ready to go in.

agatti commented:
This should be it - since last time you've looked at the code I've added the proper logic reversal for the magic bytes check, the platform name comparisons in the mpy build CI step now use explicit strings (as in quoted platform names), and I've added a message to the original PR's commit explaining the picolibc workaround and providing a bit more information about the change itself.

dpgeorge reviewed (approved):
Thanks for updating!
